testsupport/assertions/object.go

```diff
@@ -0,0 +1,281 @@
+	require.True(t, found, "no ObjectNamespaceAssertion found in the assertions but one required")
+
+	t.Logf("waiting for the first object of type %T in namespace '%s' to match criteria", newObject[T](), namespace)
+
+	possibleGvks, _, err := f.cl.Scheme().ObjectKinds(newObject[T]())
+	require.NoError(t, err)
+	require.Len(t, possibleGvks, 1)
+
+	gvk := possibleGvks[0]
+
+	var returnedObject T
+
+	ft := &errorCollectingT{logger: t}
+
+	err = kwait.PollUntilContextTimeout(ctx, f.tick, f.timeout, true, func(ctx context.Context) (done bool, err error) {
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(gvk)
+		ft.errors = nil
+		if err := f.cl.List(ctx, list, client.InNamespace(namespace)); err != nil {
+			return false, err
+		}
+		for _, uobj := range list.Items {
+			uobj := uobj
+			obj, err := remarshal[T](f.cl.Scheme(), &uobj)
+			if err != nil {
+				return false, fmt.Errorf("failed to cast object with GVK %v to object %T: %w", gvk, newObject[T](), err)
+			}
+
+			f.assertions.Test(ft, obj)
+
+			if !ft.failed {
+				returnedObject = obj
+			}
+		}
+		return !ft.failed, nil
+	})
+	if err != nil {
+		sb := strings.Builder{}
+		sb.WriteString("failed to find objects (of GVK '%s') in namespace '%s' matching the criteria: %s")
+		args := []any{gvk, namespace, err.Error()}
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(gvk)
+		if err := f.cl.List(context.TODO(), list, client.InNamespace(namespace)); err != nil {
+			sb.WriteString(" and also failed to retrieve the object at all with error: %s")
+			args = append(args, err)
+		} else {
+			sb.WriteString("\nlisting the objects found in cluster with the differences from the expected state for each:")
+			for _, o := range list.Items {
+				o := o
+				obj, _ := remarshal[T](f.cl.Scheme(), &o)
+				key := client.ObjectKeyFromObject(obj)
+
+				sb.WriteRune('\n')
+				sb.WriteString("object ")
+				sb.WriteString(key.String())
+				sb.WriteString(":\nSome of the assertions failed to match the object (see output above).")
+			}
+		}
+		t.Logf(sb.String(), args...)
+	}
+
+	return returnedObject
+}
+
+func (f *Await[T]) Matching(ctx context.Context, t RequireT) T {
+	t.Helper()
+
+	name, found := findNameFromAssertions(f.assertions)
+	require.True(t, found, "ObjectNameAssertion not found in the list of assertions but one is required")
+
+	namespace, found := findNamespaceFromAssertions(f.assertions)
+	require.True(t, found, "ObjectNamespaceAssertion not found in the list of assertions but one is required")
+
+	key := client.ObjectKey{Name: name, Namespace: namespace}
+
+	t.Logf("waiting for %T with name '%s' in namespace '%s' to match additional criteria", newObject[T](), key.Name, key.Namespace)
+
+	var returnedObject T
+
+	ft := &errorCollectingT{logger: t}
+
+	err := kwait.PollUntilContextTimeout(ctx, f.tick, f.timeout, true, func(ctx context.Context) (done bool, err error) {
+		t.Helper()
+		ft.errors = nil
```

Comment:
do you know if we could replace this with either `assert.Eventually` or `assert.EventuallyWithT` ?